Description: Introduce a lightdm-guest-session-wrapper session command which MAC systems like AppArmor and SELinux can use for attaching a restrictive policy to guest sessions.
Author: Martin Pitt <martin.pitt@ubuntu.com>
Origin: Backported from trunk r1242
Bug-Ubuntu: https://launchpad.net/bugs/849027

=== modified file 'src/Makefile.am'
Index: lightdm/src/Makefile.am
===================================================================
--- lightdm.orig/src/Makefile.am	2011-10-07 11:29:02.533904703 +0200
+++ lightdm/src/Makefile.am	2011-10-07 11:29:03.417904736 +0200
@@ -79,6 +79,7 @@
 lightdm_CFLAGS = \
 	$(LIGHTDM_CFLAGS) \
 	$(WARN_CFLAGS) \
+	-DLIBEXEC_DIR=\"$(libexecdir)\" \
 	-DPKGLIBEXEC_DIR=\"$(pkglibexecdir)\" \
 	-DSBIN_DIR=\"$(sbindir)\" \
 	-DCONFIG_DIR=\"$(sysconfdir)/lightdm\" \
@@ -92,6 +93,14 @@
 	$(LIGHTDM_LIBS) \
 	-lpam
 
+libexec_PROGRAMS = lightdm-guest-session-wrapper
+
+lightdm_guest_session_wrapper_SOURCES = lightdm-guest-session-wrapper.c
+
+lightdm_guest_session_wrapper_CFLAGS = \
+	$(LIGHTDM_CFLAGS) \
+	$(WARN_CFLAGS)
+
 EXTRA_DIST = ldm-marshal.list \
 	display-manager.xml
 
Index: lightdm/src/display.c
===================================================================
--- lightdm.orig/src/display.c	2011-10-07 11:29:02.557904704 +0200
+++ lightdm/src/display.c	2011-10-07 11:29:03.421904737 +0200
@@ -440,6 +440,15 @@
         }
     }
 
+    /* for a guest session, run command through the wrapper covered by MAC */
+    if (display->priv->autologin_guest)
+    {
+        gchar *t = command;
+        command = g_strdup_printf (LIBEXEC_DIR "/lightdm-guest-session-wrapper %s", command);
+        g_debug("Guest session, running session command through wrapper: %s", command);
+        g_free (t);
+    }
+
     g_signal_emit (display, signals[CREATE_SESSION], 0, &session);
     g_return_val_if_fail (session != NULL, NULL);
 
Index: lightdm/src/lightdm-guest-session-wrapper.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ lightdm/src/lightdm-guest-session-wrapper.c	2011-10-07 11:29:03.421904737 +0200
@@ -0,0 +1,25 @@
+/* -*- Mode: C; indent-tabs-mode: nil; tab-width: 4 -*-
+ *
+ * Copyright (C) 2011 Canonical Ltd.
+ * Author: Martin Pitt <martin.pitt@ubuntu.com>
+ * 
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option) any later
+ * version. See http://www.gnu.org/copyleft/gpl.html the full text of the
+ * license.
+ */
+
+/* This is a simple wrapper which just re-execve()'s the program given as its
+ * arguments. This allows MAC systems like AppArmor or SELinux to apply a
+ * policy on this wrapper which applies to guest sessions only. */
+
+#include <unistd.h>
+
+int
+main (int argc, char *argv[], char *envp[])
+{
+    if (argc < 2)
+	return 1;
+    execve (argv[1], argv+1, envp);
+}
